### PR TITLE
Fix iOS8 crash-on-load for M63

### PR DESCRIPTION
**Pull request by michaelkirk**
If the overall diff is too noisy, please look at the individual commits. They are intended to be semantically well structured.

**Background**: Upstream WebRTC hasn't officially supported iOS8 since M58 or so. We had a previous work around which worked through M61, but broke again in M63.

The proximate culprit is that in M63 builds, MetalKit is now linked strongly.

Compare this (trimmed) output from `otool -l`

**M61 dylib loads (doesn't crash)**

```
...
1191 Load command 28                                                                                                                                  
1192           cmd LC_LOAD_WEAK_DYLIB                                                                                                                 
1193       cmdsize 80                                                                                                                                 
1194          name /System/Library/Frameworks/MetalKit.framework/MetalKit (offset 24)                                                                 
1195    time stamp 2 Wed Dec 31 16:00:02 1969                                                                                                         
1196       current version 98.3.0                                                                                                                     
1197 compatibility version 1.0.0
...
```

**M63 dylib loads (crashes on 64bit iOS8)**

```
...
1167 Load command 28                                                                                                                                  
1168           cmd LC_LOAD_DYLIB                                                                                                                      
1169       cmdsize 80                                                                                                                                 
1170          name /System/Library/Frameworks/MetalKit.framework/MetalKit (offset 24)                                                                 
1171    time stamp 2 Wed Dec 31 16:00:02 1969                                                                                                         
1172       current version 98.3.0                                                                                                                     
1173 compatibility version 1.0.0    
...
```

Note that in M61, it is a LOAD_***WEAK***_DYLIB

Note that both also contain a `LC_LOAD_DYLIB` for Metal.framework, but that's fine -  as Metal.framework is available in iOS8. It's Metal**Kit**.framework which isn't available until iOS9.

Another suspect difference is in the frameworks Info.plist

## Info.plist

**M61 Info.plist (doesn't crash)**

```
...
 37   <key>MinimumOSVersion</key>                                                                                                                     
 38   <string>8.0</string>   
...
```

**M63 Info.plist (crashes on 64bit iOS8)**

```
...
 37   <key>MinimumOSVersion</key>                                                                                                                     
 38   <string>9.0</string>   
...
```

I believe this is the (intentional) result of this commit in webrtc:

```
commit 32a11817c44bb50e4f3a69f695dcd0fc3a5bb361
Author: Oleh Prypin <oprypin@webrtc.org>
Date:   Thu Oct 5 11:11:42 2017 +0200

    Change iOS deployment target to 9.0

    8.0 is already unsupported

    Bug: webrtc:8338
    Change-Id: I708e5c90032e86c1dcbf05d0ea8e38855a510e51
    Reviewed-on: https://webrtc-review.googlesource.com/6802
    Reviewed-by: Daniela Jovanoska Petrenko <denicija@webrtc.org>
    Reviewed-by: Henrik Kjellander <kjellander@webrtc.org>
    Commit-Queue: Oleh Prypin <oprypin@webrtc.org>
    Cr-Commit-Position: refs/heads/master@{#20162}
```

Changing the minimum deployment target back to 8.0 (and fixing the resultant compiler breakages) is sufficient to change the MetalKit load back to a "weak" load, and avoids the crash.  Presumably the minimum deployment target is used to optimize weak loads into strong loads when it knows the framework should be available.

So here's the `otool -l` output from M63 with the new patches applied.

```
...
1167 Load command 28                                                                                                                                  
1168           cmd LC_LOAD_WEAK_DYLIB                                                                                                                 
1169       cmdsize 80                                                                                                                                 
1170          name /System/Library/Frameworks/MetalKit.framework/MetalKit (offset 24)                                                                 
1171    time stamp 2 Wed Dec 31 16:00:02 1969                                                                                                         
1172       current version 98.3.0                                                                                                                     
1173 compatibility version 1.0.0  
...
```

In the process of doing this work, since I was doing lots of builds, I've rounded out some of our build automation.

PTAL @charlesmchen

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+# things that are changed by our build process, at the expense of some robustness
+update_tools:
+	cd $(WEBRTC_SRC_DIR) && \
+		gclient runhooks --jobs 16
+
+build:
+	cd $(WEBRTC_SRC_DIR) && \
+		tools_webrtc/ios/build_ios_libs.sh
+
+log_build_env:
```
Comment (michaelkirk) [inline Makefile:30]: This step is now part of the default build task, and records some build environment details in a text file that lives within the WebRTC.framework which will be checked into our library repository (Signal-Carthage).

File: ios-patches/003_detect_metalkit.diff [hunk 2]
```diff
@@ -1,20 +0,0 @@
-diff --git a/sdk/objc/Framework/Classes/Metal/RTCMTLVideoView.m b/sdk/objc/Framework/Classes/Metal/RTCMTLVideoView.m
-index e36cb3ab4..012c2ecd8 100644
---- a/sdk/objc/Framework/Classes/Metal/RTCMTLVideoView.m
-+++ b/sdk/objc/Framework/Classes/Metal/RTCMTLVideoView.m
-@@ -60,7 +60,14 @@
- 
- + (BOOL)isMetalAvailable {
```
Comment (michaelkirk) [inline ios-patches/003_detect_metalkit.diff:7]: This change didn't go away, it now lives in the (renamed) patch `003_support_ios8.diff` and includes the DEPLOYMENT_TARGET changes which iOS8 also requires.

File: ios-patches/003_support_ios8.diff [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+index 4afa15d11..5876ee86e 100644
+--- a/sdk/objc/Framework/Classes/Common/helpers.mm
++++ b/sdk/objc/Framework/Classes/Common/helpers.mm
+@@ -120,7 +120,8 @@ int GetProcessorCount() {
+ #if defined(__IPHONE_9_0) && defined(__IPHONE_OS_VERSION_MAX_ALLOWED) \
+     && __IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_9_0
+ bool GetLowPowerModeEnabled() {
+-  if (isOperatingSystemAtLeastVersion(9.0)) {
++  //if (isOperatingSystemAtLeastVersion(9.0)) {
++  if (@available(iOS 9, *)) {
```
Comment (michaelkirk) [inline ios-patches/003_support_ios8.diff:24]: Even though logically equivalent to the replaced code, xcode expects you to use `@available` for checks like this, else it's static analysis produces warning which will fail the build.

File: ios-patches/003_support_ios8.diff [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+diff --git a/tools_webrtc/ios/build_ios_libs.py b/tools_webrtc/ios/build_ios_libs.py
+index 0051982e0..39cc88668 100755
+--- a/tools_webrtc/ios/build_ios_libs.py
++++ b/tools_webrtc/ios/build_ios_libs.py
+@@ -31,7 +31,7 @@ SDK_LIB_NAME = 'librtc_sdk_objc.a'
+ SDK_FRAMEWORK_NAME = 'WebRTC.framework'
+ 
+ DEFAULT_ARCHS = ENABLED_ARCHS = ['arm64', 'arm', 'x64', 'x86']
+-IOS_DEPLOYMENT_TARGET = '9.0'
++IOS_DEPLOYMENT_TARGET = '8.0'
```
Comment (michaelkirk) [inline ios-patches/003_support_ios8.diff:85]: Here's the primary new change which prevents the crash on iOS8. Pretty obvious once I found it...

File: Makefile (hunk 1 above)
Comment (charlesmchen) [reply Makefile:30]: +1

File: bin/clean_webrtc.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python
+
+# Attempts to reverse any Signal specific modifications to webrtc.
+# If our patches change, this may have to be revisted in the future.
```
Comment (charlesmchen) [inline bin/clean_webrtc.py:4]: s/revisted/revisited/

File: bin/print_build_env.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+def get_build_details():
+    template = Template("""## WebRTC Build Details
+
+To track down potential future issues, we log some of our build environment details.
+
+git branch:
+$git_branch
+
+xcode-select -p:
+$xcode_path
```
Comment (charlesmchen) [inline bin/print_build_env.py:41]: Don't you want to capture the xcode version too?

File: bin/print_build_env.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        git_branch = [line.replace("* ", "") for line in git_branch_output.split("\n") if re.search("^\*", line)][0]
+        return git_branch
+
+def get_build_details():
+    template = Template("""## WebRTC Build Details
+
+To track down potential future issues, we log some of our build environment details.
+
+git branch:
+$git_branch
```
Comment (charlesmchen) [inline bin/print_build_env.py:38]: git commit too?

File: bin/print_build_env.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+xcode-select -p:
+$xcode_path
+
+hostname:
+$hostname
+
+gcc -v:
+$gcc_version
+""")
```
Comment (charlesmchen) [inline bin/print_build_env.py:48]: os x version too?

**Review by charlesmchen — APPROVED** [review]
LGTM

File: bin/print_build_env.py (hunk 6 above)
Comment (michaelkirk) [reply bin/print_build_env.py:41]: Done!

File: bin/print_build_env.py (hunk 7 above)
Comment (michaelkirk) [reply bin/print_build_env.py:38]: Done! 

I'm now capturing two sha's, one for the upstream webrtc repo, and one for the build tools (this repo)

File: bin/print_build_env.py (hunk 8 above)
Comment (michaelkirk) [reply bin/print_build_env.py:48]: done via sw_vers. Final output looks like this:
```
## WebRTC Build Details

To track down potential future issues, we log some of our build environment details.

webrtc git branch:
(HEAD detached at branch-heads/63)

webrtc git sha:
dc5904661c4c8b357e19595d80b2881db3f4a71b

build_script git sha:
6847d7ada965e930f57880274ef3b1abe2bf01dd

xcodebuild -version:
Xcode 9.2
Build version 9C40b

xcode-select -p:
/Applications/Xcode9.2.app/Contents/Developer

gcc -v:
Configured with: --prefix=/Applications/Xcode9.2.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 9.0.0 (clang-900.0.39.2)
Target: x86_64-apple-darwin16.7.0
Thread model: posix
InstalledDir: /Applications/Xcode9.2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

osx_version_details:
ProductName:    Mac OS X
ProductVersion: 10.12.6
BuildVersion:   16G1114

hostname:
ash.local
```

File: bin/clean_webrtc.py (hunk 5 above)
Comment (michaelkirk) [reply bin/clean_webrtc.py:4]: Thanks!

**Comment by michaelkirk** [comment]
In https://github.com/signalapp/signal-webrtc-ios/commit/2a793074f4cfd6f9ed121724b52963de68134372

**Pull request closed without merging**